src/orange/detail/BIHPartitioner.cc

```diff
@@ -0,0 +1,175 @@
+/*!
+ * Sort and uniquify a vector.
+ */
+void sort_and_uniquify(std::vector<celeritas::real_type>& vec,
+                       celeritas::real_type tol)
+{
+    std::sort(vec.begin(), vec.end());
+
+    auto last = std::unique(vec.begin(),
+                            vec.end(),
+                            [&](celeritas::real_type a, celeritas::real_type b) {
+                                return std::abs(a - b) < tol;
+                            });
+
+    vec.erase(last, vec.end());
+}
+}  // namespace
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from vector of bounding boxes and respective centers.
+ */
+BIHPartitioner::BIHPartitioner(VecBBox* bboxes, VecReal3* centers)
+    : bboxes_(bboxes), centers_(centers)
+{
+    CELER_EXPECT(!bboxes_->empty());
+    CELER_EXPECT(bboxes_->size() == centers_->size());
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Determine if a set of bounding boxes can be partitioned
+ */
+bool BIHPartitioner::is_partitionable(VecIndices const& indices) const
+{
+    CELER_EXPECT(!indices.empty());
+
+    auto centers = this->axes_centers(indices);
+    return std::any_of(centers.begin(),
+                       centers.end(),
+                       [](std::vector<real_type> const& centers) {
+                           return centers.size() > 1;
+                       });
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Find a suitable partition for the given bounding boxes.
+ */
+BIHPartitioner::Partition
+BIHPartitioner::operator()(VecIndices const& indices) const
+{
+    CELER_EXPECT(!indices.empty());
+
+    Partition best_partition;
+    real_type best_cost = std::numeric_limits<real_type>::infinity();
+
+    auto axes_centers = this->axes_centers(indices);
+
+    for (auto axis : range(Axis::size_))
+    {
+        auto ax = to_int(axis);
+
+        for (auto i : range(axes_centers[ax].size() - 1))
+        {
+            Partition p;
+            p.axis = axis;
+            p.location = (axes_centers[ax][i] + axes_centers[ax][i + 1]) / 2;
+            this->apply_partition(indices, p);
+            auto cost = this->calc_cost(p);
+
+            if (cost < best_cost)
+            {
+                best_partition = p;
+                best_cost = cost;
+            }
+        }
+    }
+
+    CELER_VALIDATE(best_partition, << "calculated partition not valid");
+
+    return best_partition;
+}
+
+//---------------------------------------------------------------------------//
+// HELPER FUNCTIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Create sorted and uniquified X, Y, Z values of bbox centers.
+ */
+BIHPartitioner::AxesCenters
+BIHPartitioner::axes_centers(VecIndices const& indices) const
```

Comment:
Maybe
```suggestion
BIHPartitioner::calc_axes_centers(VecIndices const& indices) const
```
to indicate it's expensive?